File: packages/detached_membrane_sdk/detached_membrane_sdk/verification_bridge.py
"""Verification bridge for host validator acknowledgements."""

from __future__ import annotations

from typing import Any

VERIFICATION_SCHEMA_VERSION = "detached-membrane-verify-0_1"


def bridge_verification_result(
    *,
    packet: dict[str, Any],
    host_ack: dict[str, Any],
) -> dict[str, Any]:
    """Normalize host validator output into stable verification schema."""
    accepted = bool(host_ack.get("accepted", False))
    errors = host_ack.get("errors", [])
    if not isinstance(errors, list):
        errors = [str(errors)]

    return {
        "object": "DetachedMembraneVerificationResult",
        "schema_version": VERIFICATION_SCHEMA_VERSION,
        "accepted": accepted,
        "errors": errors,
        "packet_ref": {
            "return_id": packet.get("return_id"),
            "source_packet_id": packet.get("source_packet_id"),
            "executor_profile_id": packet.get("executor_profile_id"),
            "status": packet.get("status"),
        },
    }

File: packages/detached-membrane-governance/src/executor/verification-bridge.ts
const VERIFICATION_SCHEMA_VERSION = "detached-membrane-verify-0_1";

export function bridgeVerificationResult(
  packet: Record<string, unknown>,
  hostAck: Record<string, unknown>,
): Record<string, unknown> {
  const accepted = Boolean(hostAck.accepted);
  let errors = hostAck.errors;
  if (!Array.isArray(errors)) {
    errors = errors !== undefined ? [String(errors)] : [];
  }

  return {
    object: "DetachedMembraneVerificationResult",
    schema_version: VERIFICATION_SCHEMA_VERSION,
    accepted,
    errors,
    packet_ref: {
      return_id: packet.return_id,
      source_packet_id: packet.source_packet_id,
      executor_profile_id: packet.executor_profile_id,
      status: packet.status,
    },
  };
}
